client/src/pages/Map/MapboxManager.js

```diff
@@ -0,0 +1,85 @@
+import { REGION_TYPES } from '../../util/constants';
+
+class MapboxManager {
```

Comment:
What did you have in mind by "closures/factory functions"?

React components are recommended to be written with functions, but this class isn't a react component. It can be written with function closures, but that's basically what classes are in javascript anyway - syntactic sugar for stateful logic. I like the simplicity of this

Reply:
Yeah, I considered doing something similar to your example, but I am intentionally avoiding this. 

1. The first thing that's happening here is hiding a context and access to the context within the reusable `useMapbox` hook, which I know is a paradigm we use elsewhere in the code (like in `useNewTree`). I'm intentionally straying away from this, because I feel that it overcomplicates the code by creating a context and hiding it in this hook. It's not always clear what the API is (maybe typescript would help with this), so I'm forced to actually read the hook code. It takes me longer to understand what it's doing when all we're really doing is accessing a react context. I think directly accessing context within components is clearer IMO.

2. The other part is that the example rewrites `MapboxManager` as a React component, which I don't think is right. I don't think every piece of code in our front-end should be written in React as not every piece of logic is tied to rendering a UI component. Making MapboxManager think about react lifecycles overcomplicates things. 

For example, in the example you wrote, there is a pitfall embedded in the code from not using `useMemo` to memoize getCenter, setCenter, addMarkerRefToMap. On every rerender, these functions are being recreated, which is a negligible perf hit, but more importantly, this can cause breaking bugs that are hard to track down when using `debounce` or `throttle` since the functions are not static. 

Another example is https://github.com/waterthetrees/wtt_front/pull/668. It's actually quite more difficult to write this function if `MapboxManager` was a react component since we would need to use `useRef` as an escape hatch to avoid recreating the promise on each rerender. 

In general, React tries to keep components as idempotent as possible (which is why it double-renders in test envs) and uses `useEffect` hooks as an escape hatch for side effects. The problem I see when working with our map code is that there's a lot of brittle code related to our map side effects. Examples that I've identified in the past are: code that breaks down when we reduce the number of rerenders, code that is intended to be executed just once either executes on each rerender or doesn't execute at all (executes but before async requests have given it good data), which are both kinda tricky and unintuitive issues.

Reply:
Here's a chatgpt excerpt that actually summarizes my thoughts in a slightly different way

> In a React project, it is not necessary to make every helper function a React component. Helper functions that do not require any UI rendering should be written as normal JavaScript functions, and can be imported and used as needed in React components.

Creating unnecessary React components can add unnecessary overhead to the project and increase its complexity. Additionally, using React components for non-UI related logic can lead to confusion for other developers who may expect them to have rendering behavior.

It is generally considered best practice to separate UI-related code and business logic. UI-related code should be written in React components, while business logic should be written in regular JavaScript functions or classes.

In summary, it is not necessary to create React components for business logic helper functions that do not render anything, and it is generally best practice to separate UI-related code and business logic.

Reply:
Okay, well, the PR itself is ready, but there were merge conflicts with main. Lmk when this is approved and I can do another round of manual testing to ensure I resolved the merge conflicts correctly - don't want to waste time on unnecessary testing cycles.